server/athenian/api/experiments/preloading/cache.py

```diff
@@ -0,0 +1,259 @@
+        self._categorical_cols = categorical_cols or []
+        self._identifier_cols = identifier_cols or []
+        self._filtering_clause = filtering_clause
+        self._db = db
+        self._debug_memory = debug_memory
+        self._mem = {}
+        self._df = None
+
+    @property
+    def id_(self) -> str:
+        """Return the id of the `CachedDataFrame`."""
+        return self._id
+
+    async def load(self) -> "CachedDataFrame":
+        """Load the DataFrame from the database."""
+        query = select(self._cols)
+        if self._filtering_clause:
+            query = query.where(self._filtering_clause)
+
+        res = await read_sql_query(query, self._db, self._cols)
+        df = pd.DataFrame(res)
+        if self._debug_memory:
+            size = df.memory_usage(index=True, deep=True)
+            self._mem["raw"] = {"series": size, "total": sum(size)}
+
+        self._df = self._squeeze(df)
+        if self._debug_memory:
+            size = self._df.memory_usage(index=True, deep=True)
+            self._mem["processed"] = {"series": size, "total": sum(size)}
+            self._mem["percentage"] = {
+                "series": self._mem["processed"]["series"]
+                / self._mem["raw"]["series"]
+                * 100,
+                "total": self._mem["processed"]["total"]
+                / self._mem["raw"]["total"]
+                * 100,
+            }
+
+        return self
+
+    async def refresh(self) -> "CachedDataFrame":
+        """Refresh the DataFrame from the database."""
+        if self._df is None:
+            raise ValueError("DataFrame not loaded")
+
+        return await self.load()
+
+    def memory_usage(
+        self, total: bool = False, human: bool = False,
+    ) -> Union[pd.Series, int, str]:
+        """Return information about the memory usage."""
+        s = self._df.memory_usage(index=True, deep=True)
+        if total:
+            size = sum(s)
+            return size if not human else _humanize_size(size)
+        else:
+            return s if not human else s.apply(_humanize_size)
+
+    def _squeeze(self, df: pd.DataFrame) -> pd.DataFrame:
+        def _cast_category(df, cols):
+            for c in cols:
+                df[c.name] = df[c.name].astype("category")
+            return df
+
+        def _cast_timestamp(df, cols):
+            for c in cols:
+                df[c.name] = (df[c.name].astype("int64") / 10 ** 9).astype("int32")
+                df.loc[df[c.name] < 0, c.name] = 0
+            return df
+
+        def _cast_identifiers(df, cols):
+            for c in cols:
+                max_length = df[c.name].apply(len).max()
+                df[c.name] = df[c.name].astype(f"|S{max_length}")
+            return df
+
+        return (
+            df.pipe(_cast_category, self._categorical_cols)
+            .pipe(_cast_timestamp, self._timestamp_cols)
+            .pipe(_cast_identifiers, self._identifier_cols)
+        )
+
+    def __repr__(self):
```

Comment:
```suggestion
    def __str__(self):
```
I bet you meant `__str__` here.

Reply:
Actually I meant `__repr__`. I know that `__repr__` should be a "unique" representation of the object, but I took inspiration from [how `pandas` does it](https://github.com/pandas-dev/pandas/blob/9be0b8224622bc340333f05dd62be2d02996945a/pandas/core/frame.py#L922). Basically it was simply because it was handier when playing in ipython, but it's fine as well for me to have `__str__` instead.

Reply:
Yeah, pandas is not an example of great idiomatic Python haha

Reply:
haha I'll keep it as `__str__` 👍